### PR TITLE
Support -D for `jag container install`

### PR DESCRIPTION
**Pull request by kasperl**

### CODE REVIEW

File: src/jaguar.toit [hunk 1]
```diff
@@ -168,77 +206,58 @@ install_image image_size/int reader/reader.Reader defines/Map -> none:
...
-  name/string := defines["container.name"]
-  with_timeout --ms=60_000: install_mutex.do:
-    if id := registry_.uninstall name:
-      logger.info "container $id uninstalled ('$name')"
+  // Write the image into flash.
+  image := flash_image image_size reader null defines
+
+  // We start the container from a separate task to allow the HTTP server
+  // to continue operating. This also means that the container running
+  // isn't covered by the flashing mutex or associated timeout.
```
Comment (floitsch) [inline src/jaguar.toit:225]: ?
```suggestion
  // isn't covered by the flashing mutex or associated timeout that we are currently in.
```

File: src/jaguar.toit [hunk 2]
```diff
@@ -447,11 +466,18 @@ serve_incoming_requests socket/tcp.ServerSocket id/uuid.Uuid name/string address
       writer.write
           json.encode registry_.entries
 
+    // Handle installing containers.
+    else if path == "/install" and request.method == "PUT":
+      container_name ::= headers.single HEADER_CONTAINER_NAME
+      defines ::= extract_defines headers
+      install_image request.content_length request.body container_name defines
+      writer.write
+          json.encode {"status": "OK"}
```
Comment (floitsch) [inline src/jaguar.toit:475]: Could be interesting to have an `OK_` static, but maybe not worth it.

**Review by floitsch — APPROVED** [review]
LGTM.

File: src/jaguar.toit [hunk 3]
```diff
@@ -154,11 +162,41 @@ run id/uuid.Uuid name/string port/int:
...
+        writer.write data
+      logger.debug "installing container image with $image_size bytes -> wrote $written_size bytes"
+      writer.commit --data=(name != null ? JAGUAR_INSTALLED_MAGIC : 0)
+    return image
+  unreachable
+
+run_image image/uuid.Uuid cause/string name/string? defines/Map -> containers.Container:
+  nick := name ? "container '$name'" : "program $image"
+  suffix := defines.is_empty ? "" : " with $defines"
+  defines.filter --in_place: not it.starts_with "jag."
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/jaguar.toit:182]: Avoided the modifications.

File: src/jaguar.toit (hunk 2 above)
Comment (kasperl) [reply src/jaguar.toit:475]: Will look at that in another PR.

File: src/jaguar.toit [hunk 4]
```diff
@@ -52,7 +58,9 @@ registry_ / ContainerRegistry ::= ContainerRegistry
 
 main arguments:
   try:
-    catch --trace: registry_.start_installed
+    catch --trace:
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/jaguar.toit:61]: Done.

File: src/container_registry.toit [hunk 5]
```diff
@@ -82,16 +86,20 @@ class ContainerRegistry:
       // with the correct structure, so we guard the access to the
       // individual entries and treat malformed ones as non-existing.
       name/string? := null
-      catch: name = entries.get "$image.id"
+      defines/Map? := null
+      entry := entries.get "$image.id"
+      catch: name = entry[0]
+      catch: defines = entry[1]
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/container_registry.toit:92]: I'm slightly worried that entry might be a non-map.

File: src/container_registry.toit [hunk 6]
```diff
@@ -15,23 +15,25 @@ jaguar_ / uuid.Uuid ::= containers.current
...
-  loaded_     / bool := false
-  id_by_name_ / Map ::= {:}  // Map<string, uuid.Uuid>
-  name_by_id_ / Map ::= {:}  // Map<uuid.Uuid, string>
+  loaded_        / bool := false
+  id_by_name_    / Map ::= {:}  // Map<string, uuid.Uuid>
+  name_by_id_    / Map ::= {:}  // Map<uuid.Uuid, string>
+  defines_by_id_ / Map ::= {:}  // Map<uuid.Uuid, Map>
 
-  entries -> Map:
+  entries --defines/bool=false -> Map:
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/container_registry.toit:23]: Yeah. I'll try to streamline this a bit in another PR.

**Pull request merged**